test/test_foreach.py

```diff
@@ -77,23 +139,314 @@ def _get_test_data(self, device, dtype, N):
+                    self.assertEqual(expected, inputs[0], atol=1.e-3, rtol=self.dtype_precisions[dtype][0])
+                else:
+                    self.assertEqual(expected, inputs[0])
+
+    def _test_binary_op_tensorlists(self, device, dtype, opinfo, N, is_fastpath):
+        op, ref, inplace_op, inplace_ref = self._get_funcs(
+            opinfo, dtype=dtype, scalars=None, n_tensors=N, is_fastpath=is_fastpath)
+        inputs = [
+            opinfo.sample_inputs(device, dtype, N, noncontiguous=not is_fastpath),
+            opinfo.sample_inputs(device, dtype, N, noncontiguous=not is_fastpath),
+        ]
+        self._regular_binary_test(dtype, op, ref, inputs, is_fastpath)
+        self._inplace_binary_test(dtype, inplace_op, inplace_ref, inputs, is_fastpath)
+        if opinfo.supports_alpha_param:
+            alpha = 3 if dtype in torch.testing.get_all_int_dtypes() else 3.14
+            self._regular_binary_test(dtype, op, ref, inputs, is_fastpath, alpha=alpha)
+            self._inplace_binary_test(dtype, inplace_op, inplace_ref, inputs, is_fastpath, alpha=alpha)
+
+    @ops(foreach_binary_op_db)
+    def test_binary_op_tensorlists_fastpath(self, device, dtype, op):
         for N in N_values:
-            tensors1 = self._get_test_data(device, dtype, N)
-            tensors2 = self._get_test_data(device, dtype, N)
+            self._test_binary_op_tensorlists(device, dtype, op, N, True)
 
-            # Mimics cuda kernel dtype flow.  With fp16/bf16 input, runs in fp32 and casts output back to fp16/bf16.
-            control_dtype = torch.float32 if (self.device_type == 'cuda' and
-                                              (dtype is torch.float16 or dtype is torch.bfloat16)) else dtype
-            expected = [torch_op(tensors1[i].to(dtype=control_dtype),
-                                 tensors2[i].to(dtype=control_dtype)).to(dtype=dtype) for i in range(N)]
-            res = foreach_op(tensors1, tensors2)
-            foreach_op_(tensors1, tensors2)
-            self.assertEqual(res, tensors1)
-            if (dtype is torch.float16 or dtype is torch.bfloat16) and TEST_WITH_ROCM:
-                self.assertEqual(tensors1, expected, atol=1.e-3, rtol=self.dtype_precisions[dtype][0])
-            else:
-                self.assertEqual(tensors1, expected)
+    @dtypes(*torch.testing.get_all_dtypes())
+    @ops(foreach_binary_op_db)
+    def test_binary_op_tensorlists_slowpath(self, device, dtype, op):
+        for N in N_values:
+            self._test_binary_op_tensorlists(device, dtype, op, N, False)
+
+    def _test_binary_op_scalar(self, device, dtype, opinfo, N, scalar, is_fastpath):
+        op, ref, inplace_op, inplace_ref = self._get_funcs(
+            opinfo, dtype=dtype, scalars=scalar, n_tensors=N, is_fastpath=is_fastpath)
+        inputs = [opinfo.sample_inputs(device, dtype, N, noncontiguous=not is_fastpath), scalar]
+        inputs2 = [scalar for _ in range(N)]
```

Comment:
wait, this is not `scalar`, this is scalar list, you op will be `torch._for_each_add([x,x], [1,1])`, whereas what you need to test is `torch._for_each_add([x,x], 1)`. As written, this test is not at all different from `_test_binary_op_scalarlist` below.